blog: SLSA end-to-end (#1458)
slsa-framework/slsa: docs/_posts/2025-07-21-slsa-e2e.md

---
title: Request for examples: SLSA End-to-End
author: Andrew McNamara, Tom Hennen
is_guest_post: false
---

## Goal

This document outlines a request for examples (RFE) for an end-to-end
implementation of the Supply-chain Levels for Software Artifacts
([SLSA](slsa.dev)) framework. The goal is to create a comprehensive
demonstration of how SLSA can be used to secure the software supply chain, from
source code to end-user consumption. These implementations will serve as a
reference for the community, showcasing best practices and providing a clear
adoption path for organizations looking to improve their software supply chain
security.

We’re looking for examples in the form of blog posts that document the entire
end-to-end story with working implementations that people can try for
themselves.

## Timeline

There’s no specific deadline, but a reasonable goal would be to have example
implementations published by September 30th, 2025.

## Stages

> For the purposes of the end-to-end story we’ve broken the SDLC into 5 stages

Reviewer: Mostly as an observation, this stage model combines verify/publication phase for "source" while it's separate for "build"

* source
  * verify
  * publish
* build & test
  * verify 
  * publish
* use
  * verify
  * profit!

Where the "verify" parts are repeated before starting each new phase.

which we’d expect an end-to-end implementation to cover: Source, Build & Test,
Verification, Publication, and Use.

### Source

> The source stage covers applying SLSA protections to source code leveraging the

Reviewer 2: Perhaps a nitpick, but I think we should be clear that SLSA isn't a security mechanism per se.

```suggestion
The source stage covers applying SLSA requirements to source code management leveraging the
```

SLSA Source Track
[proposed in SLSA 1.2 RC1](/spec/v1.2-rc1/tracks##source-track).

Implementations should:

-   **Protect a source repo with the SLSA Source Track:** Implement SLSA Source
  Level 3 for source control, including measures like branch protection,
  mandatory code review, and protection against tampering with the source code
  history. Use existing implementations if desired.
    -   **Record Source Provenance:** Generate and store SLSA source provenance
    attestations for all source code, capturing information about the author,
    committer, and the specific commit hash.

>     -   **Issue Source VSA:** Generate a Verifiable Source Attestation (VSA) for the

Reviewer: I belive this should be 
```suggestion
    -   **Issue Source VSA:** Generate a Verification Summary Attestation (VSA) for the
```

right? https://slsa.dev/spec/v1.2-rc1/verification_summary

    source code, which cryptographically guarantees the integrity and provenance
    of the code.

<p style="text-align: right">
Potential tools: <a href="https://github.com/slsa-framework/slsa-source-poc">slsa-source-poc</a>, <a href="https://gittuf.dev">gittuf</a></p>

> ### Build & Test

Reviewer 2: +1 to this!

> The build & test stage covers applying SLSA protections to the way software is

Reviewer 2: As above
```suggestion
The build & test stage covers applying SLSA requirements to the way software is
```

built and how the dependencies used in a build are managed. It leverages the
[SLSA Build track](/spec/v1.1/), and the draft
[Build Environment](/spec/draft/#build-environment-track) and
[Dependency](/spec/draft/#dependency-track) tracks.

Implementations should:

-   **Generate Build Provenance:** Generate SLSA build Level 3-compliant build
  provenance, which includes information about the build environment, the
  specific build steps, and the inputs and outputs of the build process.
-   **Generate Build Environment attestations**: (Optional) Generate SLSA BuildEnv

>   L2-compliant BuildEnv attestations, which includes integrity information about
>   the build environment at the start of the build.

Reviewer 2: I'm ok with leaving this step as optional, but I'm a bit confused as to why we're limiting the BuildEnv track to L2, not L3, given that it is possible to implement L3 today. I suggest adding a line to that effect.
```suggestion
  L2-compliant BuildEnv attestations, which includes integrity information about
  the build environment at the start of the build. If trusted hardware is available,
  additionally generate L3-compliant BuildEnv attestations.
```

-   **Generate an SBOM:** (Optional) Generate a Software Bill of Materials (SBOM)
  in a standard format (e.g., SPDX, CycloneDX) that lists all the components and
  dependencies of the built artifact.

> -   **Collect source VSAs for all used source repos (as listed in resolvedDendencies):** For each source,

Reviewer: It seems like we should verify VSAs for all dependencies, not just source, right?

### Verification

The verification stage covers the point at which a built software artifact is
[verified against expectations](/spec/draft/verifying-artifacts).
This will go beyond the basics identified in verifying build artifacts and
include verifying Source, Build Environment, and Dependency information.

Implementations should:

-   **Allow software producers to define expectations used during verification
  (i.e. a policy):** Provide a mechanism for software producers to define a
  security policy that specifies the requirements for a successful verification.
-   **Verification:** Implement a verification process that checks the following:
    -   **Check build provenance:**

>         -   **Comes from the expected builder:** Verify that the build was performed
>       by a trusted and authorized builder.

Reviewer 2: I'm not sure if this is meant to include checking the BuildEnv attestation. If not, I would add another sub-bullet:

```suggestion
        -   **Comes from the expected builder:** Verify that the build was performed
      by a trusted and authorized builder.
      -    **Build was run in the expected build environment (optional):** Verify that
      the build ran in a good known build environment as advertised by the trusted
      builder.
```

        -   **From the expected build entrypoint:** Verify that the build was
      initiated from the correct and expected entry point.
        -   **Each source dependency has:**
            -   **A valid source VSA:** Verify the integrity and provenance of each
        source dependency by
        [checking its VSA](/spec/v1.2-rc1/verifying-source#how-to-verify-slsa-a-source-revision).
            -   **(optionally) has a desired source level:** Verify that each source
        dependency meets the minimum SLSA source level defined in the security
        policy.
    -   **Check SBOM:** Check that an SBOM from the expected tooling exists.
    -   **Check vulnerability attestation and triage:** Verify that all identified
    vulnerabilities have been triaged and that the remediation status is
    acceptable according to the security policy.
-   **Issue VSA including:**
    -   **SLSA build level:** The VSA should include the SLSA build level that the

>     artifact has achieved.

Reviewer 2: As above, the VSA should also optionally include the BuildEnv level at which the artifact was built.

-   **Support for attestations from different producers:** The implementation may
  be able to consume and verify attestations from different producers, such as
  different builders, source control systems, and SBOM generators.

## Submission

Submissions should document:

-   **How they work:** Explains the end-to-end workflow of the implementation.
-   **How they leverage existing standards:** Note which standards or existing
  projects are used in the submission.
-   **How can a user adopt this workflow themselves:** A step-by-step guide that
  allows users to set up and use the implementation in their own environment.
  This can link to existing documentation of any technology used.
-   **The policy management process for the software producer:** A document that
  explains how to create, manage, and enforce security policies.
-   **Gaps and future areas of improvement:** A document that identifies the
  current limitations of the implementation and suggests areas for future
  improvement.
-   **A working example:** A complete, working example of the implementation.

To submit your example end-to-end implementations please either send a PR

> creating the blog post yourself

Reviewer 3: Not really important but a nit to provide a little bit more clarity:

Until I read all the way to the end I it became clear that the examples are to be published as blog posts in this line which just takes it as a given. Perhaps this should be clarified in the intro paragraphs. Coming from an implementer's point of view I was wondering about the submission medium and shape all the way until the end.

Author: Thanks, that's a good point.  Updated the objective section and clarified.
